### ISSUE TITLE
Allow debugging multiple projects in one debug session

### ISSUE THREAD

**Issue by weinand** [OmniSharp/omnisharp-vscode#460]
_From @matthewrk on June 23, 2016 13:22_

This problem specifically affects dot net core development via vscode. I'm not entirely sure if its a core problem or a vscode problem. Apologies if this issue is misplaced.

The global.json file allows you to define sub projects, similar to a solution in vs.net. A very common setup for a .net project is a web app with several class libraries. This is achievable in vscode and everything works brilliantly until you attempt to debug.

It doesn't seem possible in vscode currently to debug a web app under this setup (problems locating source code due to limitations in conf files), or to place breakpoints within the web app itself as well as class libraries to follow execution when debugging.

This seems like a pretty critical requirement for any mid-large sized project, and is a hurdle to being able to migrate to using vscode for projects (on a mac, in my case) from vs.net.

I have been discussing this problem on stack overflow looking for possible solutions:
http://stackoverflow.com/questions/37376415/debugging-in-vscode-with-multiple-dotnet-core-projects-under-one-solution 

And also in the comments here on a blog post about setting up multiple projects under one roof in vscode: 
http://forums.dotnetfoundation.org/t/referencing-another-project-in-net-core/1298/9

Where you can find more information on my attempts to achieve this functionality should it be helpful.

_Copied from original issue: Microsoft/vscode#8052_

**Comment by weinand** [comment]
@matthewrk just a tipp; you can start more than one debug session if you open new VS Code windows on subfolders. So if you you have a subfolder 'server' inside your main project, you can have one debug session on the main project folder and one on the subfolder. Yes, this is a limitation but we are working on lifting that.

**Comment by weinand** [comment]
_From @matthewrk on June 24, 2016 9:23_

@weinand as far as I understand that would work if you have two executable apps, but in this case I have one web app and several class libraries, e.g. data access, services.

**Comment by weinand** [comment]
@matthewrk aha, your problem is not about multiple debug sessions in one project but about multiple projects in one debug session. This is something that the dot net extension has to solve.

**Comment by gregg-miskelly** [comment]
@matthewrk what you are doing should work fine - the debugger debug's processes, not projects.

Was the dll built on the same machine as you are running it and from the same folder? If so, the debugger should be able to open source files because the PDB will contain the full path to where the fields were built. If not, you can use the [sourceFileMap](https://github.com/OmniSharp/omnisharp-vscode/blob/master/debugger.md#source-file-map) launch.json option to tell the debugger where to look. VS Code doesn't support any sort of UI to prompt for the source location, so this needs to be done ahead of time.

**Comment by matthewrk** [comment]
@gregg-miskelly yes it was all built on the same machine. I am attempting to use the following structure:

`
MySolution (I open vscode at this level)
    global.json
    MyWebApp/
        .vscode
            launch.json
            tasks.json
    MyDataLibrary/
    MyServices/
`

Which is basically the standard structure used in vs.net right, maybe my overall approach is wrong. It's important though that the solution should be able to contain multiple executable projects, plus the shared libraries.

At the moment I'm not sure the debugger will even run in that structure. I'll have a go at setting up a new project to confirm.

**Comment by gregg-miskelly** [comment]
The debug adapter itself doesn't know or care what the structure of your project is. VS Code does care a little -- it needs to be able to find your launch.json. But if you are successfully launching then clearly it is happy with the fact that your .vscode folder isn't at the root. The C# extension has code to try and automatically create your launch.json/tasks.json for you, and that definitely cares, and currently only handles pretty trivial cases, which this is well beyond. But it sounds like you manually worked around that.

Can you better describe what you are seeing the debugger do?

One thing: You can configure your tasks.json to build in a folder other than the root. See [this example](https://raw.githubusercontent.com/wiki/OmniSharp/omnisharp-vscode/ExampleCode/tasks.json).

**Comment by matthewrk** [comment]
@gregg-miskelly Here's my current test project in full: https://github.com/matthewrk/dotnetcoresolution

I've managed to get it running with vscode's debugger and successfully debugging lines both in the web app itself and the services library. To achieve this I added a launch config for the WebApp in the root launch.json, and set the options->cwd in tasks.json to my web app's root directory. 

This gets us past the first hurdle but if I were to add a console app into that solution, the root tasks.json won't work because its currently scoped to the tasks.json if I'm correct? I feel that I need the options->cwd inside each individual task rather than for all, then I would want to rename the task from build to buildMyWebApp and set that as the preLaunchTask for my MyWebApp launch configuration, which ties it all together nicely and allows me to continue adding more tasks specific to each project. Or perhaps I'm just doing it wrong and there's a way to hook up the nested tasks/launch configs?

**Comment by ShahabGanji** [comment]
I'm not sure that [this](http://www.codeproject.com/Articles/1110660/How-To-Build-Debug-Multiple-Projects-in-VS-Code) article might help you or not.
